### PR TITLE
feat(tsdb): Option on redissnuba to switch to Redis after a certain date

### PR DESCRIPTION
**Pull request by untitaker**
This adds an option to the redissnuba tsdb backend to allow you to start
reading from Snuba only after a certain date.

It can be used like this in your `sentry.conf.py`:

```
SENTRY_TSDB = 'sentry.tsdb.redissnuba.RedisSnubaTSDB'
SENTRY_TSDB_OPTIONS = {"switchover_timestamp": 123456}  # now + 90d
```

We want to use this in getsentry/onprem to enable automatic switchover. Note that we had discussions that maybe this backend should live in the onprem repo, but I think it's easier to add an option to the backend we already use in prod.

We cannot use ServiceDelegator as that's a getsentry thing

### CODE REVIEW

**Comment by untitaker** [comment]
This is obsolete because we already have started using redissnuba, so we probably lost the onprem data we wanted to migrate. See https://github.com/getsentry/onpremise/pull/426

**Comment by BYK** [comment]
>This is obsolete because we already have started using redissnuba, so we probably lost the onprem data we wanted to migrate.

Not necessarily. We still have _a lot_ of users on 8.x and 9.x series. I propose we reopen this PR.

**Comment by untitaker** [comment]
@BYK that means we should backport this PR to those versions?

**Review by BYK — APPROVED** [review]

File: src/sentry/tsdb/redissnuba.py [hunk 1]
```diff
@@ -66,11 +67,14 @@ def dont_do_this(callargs):
 }
 
 
-def selector_func(method, callargs):
+def selector_func(method, callargs, switchover_timestamp):
     spec = method_specifications.get(method)
     if spec is None:
         return "redis"  # default backend (possibly invoke base directly instead?)
 
+    if switchover_timestamp and time.time() < switchover_timestamp:
```
[the remark this answers is not in the document]
Comment (untitaker) [reply src/sentry/tsdb/redissnuba.py:75]: Doesn't work in Python 3 so I'd avoid that

File: src/sentry/tsdb/redissnuba.py [hunk 2]
```diff
@@ -66,11 +67,14 @@ def dont_do_this(callargs):
 }
 
 
-def selector_func(method, callargs):
+def selector_func(method, callargs, switchover_timestamp):
```
[the remark this answers is not in the document]
Comment (untitaker) [reply src/sentry/tsdb/redissnuba.py:70]: you're right this is unnecessary

File: src/sentry/tsdb/redissnuba.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BYK) [reply src/sentry/tsdb/redissnuba.py]: >Doesn't work in Python 3 so I'd avoid that

Yeah, makes sense with stronger typing.

File: tests/sentry/tsdb/test_redissnuba.py [hunk 3]
```diff
@@ -27,12 +27,12 @@ def test_redissnuba_connects_to_correct_backend():
 
     for method in methods:
         for model in should_resolve_to_redis:
-            assert "redis" == selector_func(method, get_callargs(model))
+            assert "redis" == selector_func(method, get_callargs(model), None)
```
Comment (BYK) [inline tests/sentry/tsdb/test_redissnuba.py:30]: Can revert all changes to this file then I guess?

Comment (untitaker) [reply tests/sentry/tsdb/test_redissnuba.py]: oh right, I thought your suggested change went the other way

**Review by jan-auer — APPROVED** [review]
❤️  for the descriptive doc comments.

**Comment by BYK** [comment]
Merge at will @untitaker

**Pull request merged**